x-pack/plugin/eql/src/main/java/org/elasticsearch/xpack/eql/execution/sequence/TumblingWindow.java

```diff
@@ -190,18 +187,19 @@ private void baseCriterion(int baseStage, SearchResponse r, ActionListener<Paylo
             // always create an ASC window
             info = new WindowInfo(baseStage, begin, end);
 
-            log.trace("Found {}base [{}] window {}->{}", base.descending() ? "tail ": "", base.stage(), begin, end);
+            log.trace("Found {}base [{}] window {}->{}", base.descending() ? "tail " : "", base.stage(), begin, end);
 
             // update current query for the next request
             base.queryRequest().nextAfter(end);
 
-            // early until check if dealing with a TAIL sequence
             // execute UNTIL *before* matching the results
+            // but after the window has been created
+            //
             // this is needed for TAIL sequences since the base of the window
             // is called once with the DESC query, then with the ASC one
             // thus UNTIL needs to be executed before matching the second query
             // that is the ASC base of the window
-            if (earlyUntil && until != null && baseStage == 1) {
+            if (until != null && baseStage > 0) {
```

Comment:
The main change is here, where until gets called right after the first query runs but *before* the results are matched so that the until results are available for comparison.